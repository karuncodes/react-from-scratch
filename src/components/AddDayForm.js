import { Component } from 'react'
import { PropTypes } from 'prop-types'

export class AddDayForm extends Component {

    constructor(props) {
        super(props);
    }

    render() {
        const { resort, date, powder, backcountry } = this.props;
        return <form className="add-day-form">
                <label htmlFor="resort"> Resort Name</label>
                <input id="resort" type="text" defaultValue={resort} />

                <label htmlFor="date"> Date</label>
                <input id="date" type="date" defaultValue={date} />

                <label htmlFor="powder"> Powder Day</label>
                <input id="powder" type="checkbox" defaultChecked={powder} />

                <label htmlFor="backcountry"> Backcountry Day</label>
                <input id="backcountry" type="checkbox" defaultChecked={backcountry} />
            </form>
    }
}

AddDayForm.defaultProps = {
    resort: "Park Hotel",
    date: "2018-05-21",
    powder: false,
    backcountry: true
}

AddDayForm.propTypes = {
    resort: PropTypes.string.isRequired,
    date: PropTypes.string.isRequired,
    powder: PropTypes.bool.isRequired,
    backcountry: PropTypes.bool.isRequired
}