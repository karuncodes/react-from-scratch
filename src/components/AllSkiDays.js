import { Component } from 'react'
import { Route } from 'react-router-dom'
import { SkiDayList } from './SkiDayList'

export class AllSkiDays extends Component {
    constructor(props) {
        super(props);
        this.state = {
            allSkiDays: [
                {
                    resort: "Mt Everest",
                    date: new Date("3/2/2018"),
                    powder: true,
                    backcountry: false
                },
                {
                    resort: "hotel kirkwood",
                    date: new Date("4/12/2018"),
                    powder: false,
                    backcountry: false
                },
                {
                    resort: "hotel transylvania",
                    date: new Date("6/1/2018"),
                    powder: true,
                    backcountry: true
                }
            ]
        }
    }
    render() {
        return (
            <div>
                <Route exact path="/list-days" render={props => (
                        <SkiDayList days={this.state.allSkiDays}/>
                )} />
                <Route path="/list-days/:filter" render={({match}) => (
                        <SkiDayList days={this.state.allSkiDays} filter={match.params.filter}/>
                )} />
            </div>
        )
    }
}