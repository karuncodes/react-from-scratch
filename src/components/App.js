import { Component, Fragment } from 'react'
import { SkiDayCount } from './SkiDayCount'
import { SkiDayList } from './SkiDayList'
import { Whoops404 } from './Whoops404'
import { NavigationMenu } from './NavigationMenu'
import { BrowserRouter as Router, Route, Switch } from 'react-router-dom'
export class App extends Component {
    constructor(props) {
        super(props);
        this.state = {
            allSkiDays: [
                {
                    resort: "Mt Everest",
                    date: new Date("3/2/2018"),
                    powder: true,
                    backcountry: false
                },
                {
                    resort: "hotel kirkwood",
                    date: new Date("4/12/2018"),
                    powder: false,
                    backcountry: false
                },
                {
                    resort: "hotel transylvania",
                    date: new Date("6/1/2018"),
                    powder: true,
                    backcountry: true
                }
            ]
        }
    }

    // countDays (filter) {
    //     const { allSkiDays } = this.state
    //     return allSkiDays.filter(
    //         (day) => (filter) ? day[filter] : day ).length
    // }

    render() {
        return  <Fragment>
                    <Switch>
                        <Route exact path="/" render={props => (
                                <SkiDayCount total= {25} />
                        )} />
                        <Route exact path="/list-days" render={props => (
                                <SkiDayList days={this.state.allSkiDays}/>
                        )} />
                        <Route path="/list-days/:filter" render={({match}) => (
                                <SkiDayList days={this.state.allSkiDays} filter={match.params.filter}/>
                        )} />
                        <Route component={Whoops404} />
                    </Switch>
                    <NavigationMenu />
                </Fragment>
    }
}