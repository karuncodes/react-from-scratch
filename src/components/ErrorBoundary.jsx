import React from 'react'

export class ErrorBoundary extends React.Component {
    constructor(props) {
        super(props)
        this.state = {
            hasError: false
        }
    }
    componentDidCatch (error, info) {
        this.setState({hasError: true});
        console.log("component catch", error, info);
    }
    render() {
        if(this.state.hasError) {
            return (<h2> Something went wrong </h2>)
        }
        return this.props.children;
    }
}