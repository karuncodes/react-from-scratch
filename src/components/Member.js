export const Member = ({member}) => (<div>
    <span>admin: {member.admin}</span>
    <span>name: {member.name}</span>
    <img style={{"width": '80px',
                 "height": '80px'}}
         src={member.thumbnail}>{member.name +'s'} Photo</img>
    <span>email: {member.email}</span>
    </div>
)