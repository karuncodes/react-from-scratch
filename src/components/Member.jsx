import React from 'react'
import  { FaShieldAlt }  from 'react-icons/fa'

export const Member = ({admin, name, image, email, makeAdmin, removeAdmin}) => (<div>
    <h1>{name} {(admin)? <FaShieldAlt /> :''}</h1>
    <img src={image} alt="profile picture" />
    <div> email: <a href={`mailto:${email}`}>{email}</a></div>
    {
        (!admin)?
        <a onClick={() => makeAdmin(email)}> Make Admin </a> :
        <a onClick={() => removeAdmin(email)}> Remove Admin </a>

    }
    </div>
)