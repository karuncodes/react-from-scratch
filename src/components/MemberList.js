import { Component } from 'react'
import { Member } from './Member'
import fetch from 'isomorphic-fetch'

export class MemberList extends Component {
    constructor(props) {
        super(props);
        this.state = {
            members: [],
            loading: false,
            administrators: []
        }
        this.makeAdmin = this.makeAdmin.bind(this);
        this.removeAdmin = this.removeAdmin.bind(this);
    }

    makeAdmin(email) {
        const administrators = [
            ...this.state.administrators,
            email
        ]
        this.setState({administrators});
    }

    removeAdmin(email) {
        const administrators = this.state.administrators.filter(em => em !== email)
        this.setState({administrators});
    }
    componentDidMount() {
        this.setState({ loading: true });
        fetch("https://api.randomuser.me/?nat=US&results=12")
        .then(response => response.json())
        .then(json => json.results )
        .then ( members => {
            this.setState(
                {
                    members,
                    loading: false
                }
            )
        })
    }
    render() {

        const { members, loading } = this.state;
        return <div className="member-list">
            {(loading)?
                <span> loading </span> :
                <span> {members.length} members </span> }
            <h1>Society Names </h1>
            {(members.length) ? members.map((member, i) => {
                return <Member key={i}
                               admin={this.state.administrators.some(em => em === member.email)}
                               name={member.name.first + ' ' + member.name.last}
                               email={member.email}
                               image={member.picture.thumbnail}
                               makeAdmin={this.makeAdmin}
                               removeAdmin={this.removeAdmin}/>
            }):<span>Currently 0 members</span>}
        </div>
    }
}