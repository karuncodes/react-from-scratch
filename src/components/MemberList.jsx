import React from 'react'
import { Member } from './Member'
import fetch from 'isomorphic-fetch'

export class MemberList extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            members: [],
            loading: false,
            administrators: []
        }
        this.makeAdmin = this.makeAdmin.bind(this);
        this.removeAdmin = this.removeAdmin.bind(this);
        this.controller = new AbortController();
        this.signal = this.controller.signal;
    }

    makeAdmin(email) {
        const administrators = [
            ...this.state.administrators,
            email
        ]
        this.setState({administrators});
    }

    removeAdmin(email) {
        const administrators = this.state.administrators.filter(em => em !== email)
        this.setState({administrators});
    }
    componentDidMount() {
        this.setState({ loading: true });
        fetch("https://api.randomuser.me/?nat=US&results=12", {
            method: 'get',
            signal: this.signal
        })
        .then(response => response.json())
        .then(json => json.results )
        .then ( members => {
            // if component unmounted before fetch is done then,
            //it creates a warning of cannot call setState
            this.setState(
                {
                    members,
                    loading: false
                }
            )
        })
        .catch(err => {
            console.log("err",err);
        })
    }
    componentWillUnmount() {
        if(!this.signal.aborted){
            this.controller.abort();
        }
    }
    render() {

        const { members, loading } = this.state;
        return <div className="member-list">
            {(loading)?
                <span> loading </span> :
                <span> {members.length} members </span> }
            <h1>Society Names </h1>
            {(members.length) ? members.map((member, i) => {
                return <Member key={i}
                               admin={this.state.administrators.some(em => em === member.email)}
                               name={member.name.first + ' ' + member.name.last}
                               email={member.email}
                               image={member.picture.thumbnail}
                               makeAdmin={this.makeAdmin}
                               removeAdmin={this.removeAdmin}/>
            }):<span>Currently 0 members</span>}
        </div>
    }
}