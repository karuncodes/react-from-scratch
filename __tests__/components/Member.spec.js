import React from 'react'
import { Member } from '../../src/components/Member'
import renderer from 'react-test-renderer'

const member = {
    admin: true,
    name: "manofzeal",
    email: "kk@manofzeal.in",
    image: "kk.jpg"
}

it('renders member',() => {
    const tree = renderer
                    .create(<Member
                        admin={member.admin}
                        name={member.name}
                        email={member.email}
                        image={member.image}/>)
                    .toJSON();
    expect(tree).toMatchSnapshot();
})