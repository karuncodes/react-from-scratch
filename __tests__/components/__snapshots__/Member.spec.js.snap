// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders member 1`] = `
<div>
  <h1>
    manofzeal
     
    <svg
      attr={
        Object {
          "viewBox": "0 0 512 512",
        }
      }
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 512 512"
      width="1em"
    >
      <path
        d="M496 128c0 221.282-135.934 344.645-221.539 380.308a48 48 0 0 1-36.923 0C130.495 463.713 16 326.487 16 128a48 48 0 0 1 29.539-44.308l192-80a48 48 0 0 1 36.923 0l192 80A48 48 0 0 1 496 128zM256 446.313l.066.034c93.735-46.689 172.497-156.308 175.817-307.729L256 65.333v380.98z"
      />
    </svg>
  </h1>
  <img
    alt="profile picture"
    src="kk.jpg"
  />
  <div>
     email: 
    <a
      href="mailto:kk@manofzeal.in"
    >
      kk@manofzeal.in
    </a>
  </div>
  <a
    onClick={[Function]}
  >
     Remove Admin 
  </a>
</div>
`;
